packages/textfield/src/mwc-textfield-base.ts

```diff
@@ -0,0 +1,261 @@
+        ${this.outlined ? this.renderOutlined() : this.renderLabelText()}
+      </label>
+      ${(this.helper || this.charCounter) ? this.renderHelperText() : ''}
+    `;
+  }
+
+  protected renderInput() {
+    return html`
+      <input id="text-field"
+             class="mdc-text-field__input"
+             type="${this.type}"
+             .value="${this.value}"
+             ?disabled="${this.disabled}"
+             placeholder="${this.placeholder}"
+             ?required="${this.required}"
+             maxlength="${this.maxlength}"
+             @change="${this.handleInputChange}">`;
+  }
+
+  protected renderIcon(icon: String) {
+    return html`<i class="material-icons mdc-text-field__icon">${icon}</i>`;
+  }
+
+  protected renderOutlined() {
+    return html`
+      <div class="mdc-notched-outline">
+        <div class="mdc-notched-outline__leading"></div>
+        ${this.label ? html`<div class="mdc-notched-outline__notch">
+          <span class="mdc-floating-label">${this.label}</span>
+        </div>` : ''}
+        <div class="mdc-notched-outline__trailing"></div>
+      </div>`;
+  }
+
+  protected renderLabelText() {
+    return html`
+      ${this.label && !this.fullWidth ? html`<span class="mdc-floating-label">${this.label}</span>` : ''}
+      <div class="mdc-line-ripple"></div>
+    `;
+  }
+
+  protected renderHelperText() {
+    const classes = {
+      'mdc-text-field-helper-text--persistent': this.helperPersistent,
+    };
+    return html`
+      <div class="mdc-text-field-helper-line">
+        <div class="mdc-text-field-helper-text ${classMap(classes)}">${this.helper}</div>
+        ${this.charCounter ? html`<div class="mdc-text-field-character-counter"></div>` : ''}
+      </div>
+    `;
+  }
+
+  protected handleInputChange() {
+    this.value = this.formElement.value;
+  }
+
+  protected createFoundation() {
+    if (this.mdcFoundation !== undefined) {
+      this.mdcFoundation.destroy();
+    }
+    this._characterCounter = this.charCounterElement ? new MDCTextFieldCharacterCounter(this.charCounterElement) : null;
+    this.mdcFoundation = new this.mdcFoundationClass(this.createAdapter(), {
+      characterCounter: this._characterCounter ? this._characterCounter.foundation : undefined
+    });
+    this.mdcFoundation.init();
+  }
+
+  protected createAdapter(): MDCTextFieldAdapter {
+    this._floatingLabel = this.labelElement ? new MDCFloatingLabel(this.labelElement) : null;
+    this._lineRipple = this.lineRippleElement ? new MDCLineRipple(this.lineRippleElement) : null;
+    this._outline = this.outlineElement ? new MDCNotchedOutline(this.outlineElement) : null;
+    return {
+      ...addHasRemoveClass(this.mdcRoot),
+      ...this.getRootAdapterMethods(),
+      ...this.getInputAdapterMethods(),
+      ...this.getLabelAdapterMethods(),
+      ...this.getLineRippleAdapterMethods(),
+      ...this.getOutlineAdapterMethods(),
+    };
+  }
+
+  private getRootAdapterMethods() {
```

Comment:
Instead of `private`, let's use `protected` instead. We might need to call these in subclasses.